problem-4/problem-4.test.js

```diff
@@ -1,21 +1,70 @@
-const solution = () => {
+//1. 가장 익숙한 방법으로 문제를 해결해 주세요.
+{
+  /*
+const solution = (n) => {
+  let decimal = 0;
+  let exponent = 0;
+  if (n === "0" || n === "1") {
+    return parseInt(n, 10);
+  }
+  const length = n.length;
+  while (exponent <= length - 1) {
+    decimal += parseInt(n.charAt(n.length - 1), 10) * 2 ** exponent;
+
+    exponent++;
+    n = n.slice(0, -1);
+  }
+  return decimal;
+};*/
+}
+//2. 이번에는 재귀 함수로 문제를 해결해 주세요.
+{
+  /*
+const solution = (n, acc = 0, exponent = 0) => {
+  if (n.length === 1) {
+    return acc + parseInt(n, 10) * 2 ** exponent;
+  }
+  acc += parseInt(n.charAt(n.length - 1), 10) * 2 ** exponent;
+  return solution(n.slice(0, -1), acc, exponent + 1);
+};*/
+}
+
+//3. 꼬리 재귀 함수로 바꿔보세요.
+//4. 꼬리 재귀 최적화를 통해서 최적화해 보세요.
+const solution = (n) => {
+  let acc = 0;
+  let exponent = 0;
+
+  const length = n.length;
+  if (length === 1) {
+    return acc + parseInt(n, 10) * 2 ** exponent;
+  }
+
+  while (exponent <= length - 1) {
+    acc += parseInt(n.charAt(n.length - 1), 10) * 2 ** exponent;
```

Comment:
문자열 마지막 문자를 숫자로 변경한 후에 제곱을 더해주고 있는 것 같아요. 문자열 마지막이라면 index로 접근해도 좋을 것 같아요

```suggestion
    acc += parseInt(n.charAt(n.length - 1), 10) * 2 ** exponent;
```